### PR TITLE
Rails 8 upgrade

### PR DESCRIPTION
**Pull request by dcyoung-dev**
## Context

Updating to the latest version of Rails

Please review by commit 🙈 

## Changes proposed in this pull request

Ran `rails app:update` which created/edited a lot of files.

- Updates many auto-generated config files, some changes just shift the content around to make it easier for future upgrades.
- With Rails 8 the default message for ActionController::ParameterMissing
changed, it ends with `is empty or invalid:` not just `is empty:` This changes the specs and some api error responses. I don't expect people to match the error response of our API word for word so changing it should be fine 🤞🏻 
- Other small changes to the code to work with rails 8.

## Guidance to review

Go on review app and play around with applying or with the support functionality or something else 😄 

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)
- [ ] Inform data insights team due to database changes
- [x] Make sure all information from the Trello card is in here
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally
- [x] Add PR link to Trello card

### CODE REVIEW

**Comment by dcyoung-dev** [comment]
[Rails 8 upgrade](https://trello.com/c/4anozgSe/353-prepare-for-rails-8-upgrade)

File: db/schema.rb [hunk 1]
```diff
@@ -624,6 +625,26 @@
...
+  end
+
+  create_table "pghero_space_stats", force: :cascade do |t|
+    t.text "database"
+    t.text "schema"
+    t.text "relation"
+    t.bigint "size"
+    t.datetime "captured_at", precision: nil
+    t.index ["database", "captured_at"], name: "index_pghero_space_stats_on_database_and_captured_at"
+  end
```
Comment (dcyoung-dev) [inline db/schema.rb:646]: This needs to be removed/reverted

File: app/views/support_interface/tasks/confirm_cancel_applications_at_end_of_cycle.html.erb [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 <div class="govuk-grid-row">
   <div class="govuk-grid-column-two-thirds">
-    <%= form_with model: nil, url: support_interface_run_task_path('cancel_applications_at_end_of_cycle') do |f| %>
+    <%= form_with model: false, url: support_interface_run_task_path('cancel_applications_at_end_of_cycle') do |f| %>
```
[the remark this answers is not in the document]
Comment (CatalinVoineag) [reply app/views/support_interface/tasks/confirm_cancel_applications_at_end_of_cycle.html.erb:6]: Nothing, I'll remove it. Thanks

**Review by elceebee — APPROVED** [review]
I've tested it out on the review app as a candidate, support and a provider and everything seemed to work as expected. Nice work!

File: app/forms/concerns/wizard.rb [hunk 3]
```diff
@@ -48,7 +48,7 @@ def state
   end
 
   def state_excluded_attributes
-    %w[state_store errors validation_context]
+    %w[state_store errors validation_context context_for_validation]
```
Comment (avinhurry) [inline app/forms/concerns/wizard.rb:51]: Out of curiosity, what was the issue here? 🤔

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -191,6 +191,10 @@ def read_csv_from_zip(zip_file, filename)
     TestSuiteTimeMachine.reset
   end
 
+  config.after do
```
Comment (avinhurry) [inline spec/rails_helper.rb:194]: Could the failures be related to removing this?

https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/config/environments/test.rb#L1

File: app/controllers/candidate_interface/decisions_controller.rb [hunk 5]
```diff
@@ -26,7 +26,9 @@ def respond_to_offer
       end
     end
 
-    def decline_offer; end
+    def decline_offer
+      @respond_to_offer = false
```
Comment (avinhurry) [inline app/controllers/candidate_interface/decisions_controller.rb]: Why was this added?

**Review by avinhurry — COMMENTED** [review]
I took the review app for a spin and didn't find anything that isn't already in other environments 🎉 

I left a few questions and comments in line. I think this [whole commit](https://github.com/DFE-Digital/apply-for-teacher-training/pull/10054/commits/63029acfd8e86c7711dbaeeea5cc1b2e1e7657b1) can be dropped as the pages don't use our styling.

File: app/forms/concerns/wizard.rb (hunk 3 above)
Comment (CatalinVoineag) [reply app/forms/concerns/wizard.rb:51]: Rails 8 added a new attribute that our wizard forms need to ignore. Otherwise the wizard will complain that it doesn't know what to do with `context_for_validation`.

The wizard doesn't need this attribute to work. It will try to save it to redis and then to Active record which we don't want https://github.com/DFE-Digital/apply-for-teacher-training/blob/c74610680b464531cb4360ba34bfce16eef3d12d/app/forms/concerns/wizard.rb#L32

File: app/controllers/candidate_interface/decisions_controller.rb (hunk 5 above)
Comment (CatalinVoineag) [reply app/controllers/candidate_interface/decisions_controller.rb]: Rails doesn't allow us to pass nil to the `model` argument anymore when using form_for anymore. It allows us to pass false though.

But we can also just remove the model attribute which I did here 09a6d1b7bd33fb145527e4816e8a244072ab3c47

Comment (CatalinVoineag) [reply app/controllers/candidate_interface/decisions_controller.rb]: I think I can drop the commit that adds the false values though

File: config/initializers/new_framework_defaults_8_0.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Be sure to restart your server when you modify this file.
+#
+# This file eases your Rails 8.0 framework defaults upgrade.
+#
+# Uncomment each configuration one by one to switch to the new default.
+# Once your application is ready to run with all new defaults, you can remove
+# this file and set the `config.load_defaults` to `8.0`.
```
[the remark this answers is not in the document]
Comment (CatalinVoineag) [reply config/initializers/new_framework_defaults_8_0.rb:7]: Yep, you're right

File: spec/rails_helper.rb (hunk 4 above)
Comment (CatalinVoineag) [reply spec/rails_helper.rb:194]: No, this is something else

Comment (avinhurry) [reply spec/rails_helper.rb:194]: Hmm, I wonder why 🤔

File: public/400.html [hunk 7]
```diff
@@ -0,0 +1,114 @@
+<!doctype html>
```
[the remark this answers is not in the document]
Comment (CatalinVoineag) [reply public/400.html]: Removed them

File: app/controllers/candidate_interface/decisions_controller.rb (hunk 5 above)
Comment (avinhurry) [reply app/controllers/candidate_interface/decisions_controller.rb]: I think you might have. I went commit by commit so I probably missed it.

**Review by avinhurry — APPROVED** [review]
